packages/sources/ncfx/src/endpoint/forex.ts

```diff
@@ -1,26 +1,106 @@
-import { AdapterConfigError, Validator } from '@chainlink/ea-bootstrap'
-import { Config, ExecuteWithConfig, InputParameters } from '@chainlink/ea-bootstrap'
+import {
+  PriceEndpoint,
+  priceEndpointInputParameters,
+  PriceEndpointParams,
+} from '@chainlink/external-adapter-framework/adapter'
+import { AdapterConfig } from '@chainlink/external-adapter-framework/config'
+import { WebSocketTransport } from '@chainlink/external-adapter-framework/transports'
+import {
+  makeLogger,
+  ProviderResult,
+  SingleNumberResultResponse,
+} from '@chainlink/external-adapter-framework/util'
+import {
+  AdapterError,
+  AdapterInputError,
+} from '@chainlink/external-adapter-framework/validation/error'
+import { customSettings } from '../config'
 
-export const supportedEndpoints = ['forex']
+interface WsMessage {
+  [pair: string]: { price: number; timestamp: string }
+}
+
+export type EndpointTypes = {
+  Request: {
+    Params: PriceEndpointParams
+  }
+  Response: SingleNumberResultResponse
+  CustomSettings: typeof customSettings
+  Provider: {
+    WsMessage: WsMessage
+  }
+}
+
+const logger = makeLogger('NcfxForexEndpoint')
 
-export type TInputParameters = { base: string; quote: string }
-export const inputParameters: InputParameters<TInputParameters> = {
-  base: {
-    aliases: ['from', 'coin'],
-    description: 'The symbol of the currency to query',
-    required: true,
+export const forexTransport = new WebSocketTransport<EndpointTypes>({
+  url: (context) => context.adapterConfig.FOREX_WS_API_ENDPOINT,
+  options: (context) => {
+    const forexEncodedCreds =
+      context.adapterConfig.FOREX_WS_USERNAME && context.adapterConfig.FOREX_WS_PASSWORD
+        ? Buffer.from(
+            JSON.stringify({
+              grant_type: 'password',
+              username: context.adapterConfig.FOREX_WS_USERNAME,
+              password: context.adapterConfig.FOREX_WS_PASSWORD,
+            }),
+          ).toString('base64')
+        : ''
+    return { headers: { ncfxauth: forexEncodedCreds } }
   },
-  quote: {
-    aliases: ['to', 'market'],
-    description: 'The symbol of the currency to convert to',
-    required: true,
+  handlers: {
+    message(message): ProviderResult<EndpointTypes>[] {
+      if (Object.keys(message).length === 0) {
+        logger.debug('WS message is empty, skipping')
+        return []
+      }
+      return Object.keys(message).map((pair) => {
+        // Split forex pair with the assumption base and quote are always 3 characters long
+        const base = pair.substring(0, 3)
+        const quote = pair.substring(3)
```

Comment:
I wonder if NCFX does this on purpose, since the crypto WS returns the instrument with a `/` to delimit the base and quote asset, whereas forex doesn't. As if implying that the instrument will always be 6 chars for forex pairs.

Reply:
That was our assumption too since forex symbols are a little more standardized

Reply:
But the forex websocket functions a little differently than crypto too so could also just be another quirk.